core/src/Streamly/Internal/Data/Serialize/TH/RecHeader.hs

```diff
@@ -0,0 +1,406 @@
+            (normalB
+                 [|do (valOff, valLen :: Word32) <-
+                        deserialize
+                            $(varE currOff)
+                            $(varE arr)
+                            $(varE endOff)
+                      pure
+                          ( ($(litE fnameLit), $(varE currOff)) : $(varE acc)
+                          , valOff + w32_int valLen)|])
+            []
+
+mkDeserializeKeysDec :: Name -> Name -> SimpleDataCon -> Q [Dec]
+mkDeserializeKeysDec funcName updateFunc (SimpleDataCon cname fields) = do
+    hOff <- newName "hOff"
+    finalOff <- newName "finalOff"
+    arr <- newName "arr"
+    endOff <- newName "endOff"
+    kvEncoded <- newName "kvEncoded"
+    finalRec <- newName "finalRec"
+    let deserializeFieldExpr (Just name, ty) = do
+            let nameLit = litE (StringL (nameBase name))
+            [|case lookup $(nameLit) $(varE kvEncoded) of
+                  Nothing -> $(emptyTy name ty)
+                  Just off -> do
+                      val <- deserializeWithSize off $(varE arr) $(varE endOff)
+                      pure $ snd val|]
+        deserializeFieldExpr _ = errorUnsupported
+    method <-
+        [|do (dataOff, hlist :: CompactList (CompactList Word8)) <-
+                 deserialize $(varE hOff) $(varE arr) $(varE endOff)
+             let keys = wListToString . unCompactList <$> unCompactList hlist
+             ($(varP kvEncoded), _) <-
+                 foldlM
+                     ($(varE updateFunc) $(varE arr) $(varE endOff))
+                     ([], dataOff)
+                     keys
+             $(varP finalRec) <-
+                 $(foldl
+                       (\acc i ->
+                            [|$(acc) <*>
+                              $(deserializeFieldExpr i)|])
+                       [|pure $(conE cname)|]
+                       fields)
+             pure ($(varE finalOff), $(varE finalRec))|]
+    pure
+        [ PragmaD (InlineP funcName NoInline FunLike AllPhases)
+        , FunD
+              funcName
+              [ Clause
+                    [ VarP hOff
+                    , VarP finalOff
+                    , VarP arr
+                    , VarP endOff
+                    ]
+                    (NormalB method)
+                    []
+              ]
+        ]
+
+    where
+
+    emptyTy k ty =
+        if isMaybeType ty
+            then [|pure Nothing|]
+            else [|error $(litE (StringL (nameBase k ++ " is not found.")))|]
+
+
+mkDeserializeExpr :: Name -> Name -> Name -> SimpleDataCon -> Q Exp
+mkDeserializeExpr initialOff endOff deserializeWithKeys con = do
+    hOff <- newName "hOff"
+    [|do (hlenOff, encLen :: Word32) <-
+             deserialize $(varE initialOff) $(varE _arr) $(varE endOff)
+         ($(varP hOff), hlen1 :: Word32) <-
+             deserialize hlenOff $(varE _arr) $(varE endOff)
+         if hlen1 == $(litIntegral hlen)
+             then if $(xorCmp hval hOff _arr)
```

Comment:
Can we combine the two conditions - `if (hlen == ... && xorCmp ...)` so that we can have one else case instead of duplicating deserializeWithKeys call?